src/compiler/crystal/semantic/normalizer.cr

```diff
@@ -428,5 +428,86 @@ module Crystal
 
       super
     end
+
+    # Turn block argument unpacking to multi assigns at the beginning
+    # of a block.
+    #
+    # So this:
+    #
+    #    foo do |(x, y), z|
+    #      x + y + z
+    #    end
+    #
+    # is transformed to:
+    #
+    #    foo do |__temp_1, z|
+    #      x, y = __temp_1
+    #      x + y + z
+    #    end
+    def transform(node : Block)
+      node = super
+
+      unpacks = node.unpacks
+      return node unless unpacks
+
+      extra_expressions = [] of ASTNode
+      next_unpacks = [] of {String, Expressions}
+
+      unpacks.each do |index, expressions|
+        temp_name = program.new_temp_var_name
+        node.args[index] = Var.new(temp_name).at(node.args[index])
+
+        extra_expressions << block_unpack_multiassign(temp_name, expressions, next_unpacks)
+      end
+
+      if next_unpacks
+        while next_unpack = next_unpacks.shift?
+          var_name, expressions = next_unpack
+
+          extra_expressions << block_unpack_multiassign(var_name, expressions, next_unpacks)
+        end
+      end
+
+      body = node.body
+      case body
+      when Nop
+        node.body = Expressions.new(extra_expressions).at(node.body)
+      when Expressions
+        body.expressions = extra_expressions + body.expressions
+      else
+        new_expressions = Array(ASTNode).new(extra_expressions.size + 1)
+        new_expressions.concat(extra_expressions)
+        new_expressions.push(node.body)
+
+        node.body = Expressions.new(new_expressions).at(node.body)
```

Comment:
I believe we could append `node.body` directly and avoid an extra array:
```suggestion
        new_expressions << node.body

        node.body = Expressions.new(new_expressions).at(node.body)
```

Reply:
Oops, I blindly applied this and the code doesn't compile anymore. `new_expression` doesn't exist here if we don't define it.

Reply:
Sorry, I meant `extra_expressions`:
```suggestion
        extra_expressions << node.body
        node.body = Expressions.new(extra_expressions).at(node.body)
```
Don't you think that would work?

Reply:
Maybe! Let's try it...